ToC disables interacting with content on mobile

Malar-Natarajan opened the pull request:
Given the page {base_url}/books/college-physics/pages/preface
AND a mobile resolution

When the toc is open

Then the visible content should look disabled
AND scrolling over it should do nothing
AND clicking on a link in the content should not open the link content, instead just close the TOC

Malar-Natarajan committed:
tests attribution status in next and previous pages

Malar-Natarajan committed:
Merge branch 'master' of https://github.com/openstax/rex-web into test-attribution-new-page

Malar-Natarajan committed:
verifying next link locators

Malar-Natarajan committed:
next and previous works but with sleep

Malar-Natarajan committed:
playing around toc

Malar-Natarajan committed:
playing around toc

Malar-Natarajan committed:
playing with toc

Malar-Natarajan committed:
toc chapter click works

Malar-Natarajan committed:
page click and chapter click works

Malar-Natarajan committed:
test works but with sleep

Malar-Natarajan committed:
test works good without sleep

Malar-Natarajan committed:
test works good without sleep

Malar-Natarajan committed:
test works

Malar-Natarajan committed:
Merge branch 'master' of https://github.com/openstax/rex-web into test-attribution-new-page

Malar-Natarajan committed:
Merge branch 'master' into test-attribution-new-page

Malar-Natarajan committed:
some minor changes not affecting the test

Malar-Natarajan committed:
Merge branch 'test-attribution-new-page' of https://github.com/openstax/rex-web into test-attribution-new-page

Malar-Natarajan committed:
updated based on test_attribution review comment

Malar-Natarajan committed:
removed unused attributes

TomWoodward committed:
Merge branch 'master' into test-attribution-new-page

Malar-Natarajan committed:
next element click with title verification works as helper function but still code duplication in base.py

Malar-Natarajan committed:
Merge branch 'master' of https://github.com/openstax/rex-web into test-attribution-new-page

Malar-Natarajan committed:
Merge branch 'test-attribution-new-page' of https://github.com/openstax/rex-web into test-attribution-new-page

Malar-Natarajan committed:
Next link helper function kind of works

Malar-Natarajan committed:
helper functions created for TOC links

Malar-Natarajan committed:
removed a redundant offscreen click method in pages/base.py
Still have the original method as commented out. Can revert if my changes are wrong

Malar-Natarajan committed:
finishing C476304
removed TOC navigation from this test. It will be covered later

Malar-Natarajan committed:
Merge branch 'master' into test-attribution-new-page

Malar-Natarajan committed:
Merge branch 'test-attribution-new-page' into test-mobile-interaction

TomWoodward committed:
Merge branch 'master' into test-attribution-new-page

TomWoodward committed:
Merge branch 'master' into test-attribution-new-page

Malar-Natarajan committed:
Deleting content_item.py 

This file was needed for the TOC changes. Since the TOC part is removed from this PR, I am deleting this file

Malar-Natarajan committed:
updated review comments

Malar-Natarajan committed:
Merge branch 'master' into test-attribution-new-page

Malar-Natarajan committed:
Merge branch 'master' into test-attribution-new-page

m1yag1 committed:
Merge branch 'master' of github.com:openstax/rex-web into test-attribution-new-page

m1yag1 committed:
refactored click_and_wait to use a reference object and renamed the method

m1yag1 committed:
added offscreen_click back to the region

m1yag1 committed:
refactored content page to use updated click and check reference object on base page

m1yag1 committed:
update docstring on click_and_check helper method

Malar-Natarajan committed:
content disabled on mobile

m1yag1 committed:
Revert "update docstring on click_and_check helper method"

This reverts commit adf82be7cb08cd6a55f024bcebaf61de0d3a3986.

Malar-Natarajan committed:
working on mobile scroll lock

Malar-Natarajan committed:
updated the title to be in base.py

Malar-Natarajan committed:
merging from attribution

Malar-Natarajan committed:
mobile only not working

Malar-Natarajan committed:
test to verify content page interaction when TOC is open on mobile resolution

TomWoodward committed:
Merge branch 'master' into test-mobile-interaction

TomWoodward commented on pytest-selenium/pages/content.py:
```diff
@@ -36,6 +51,31 @@ def sidebar(self):
         return self.Attribution(self)
 
+    @property
+    def section_url_within_attribution(self):
```
is this unused?

Malar-Natarajan replied:
@TomWoodward this is used in test_section_url_in_citation_text_shows_url_for_current_page

TomWoodward commented on pytest-selenium/pages/content.py:
```diff
@@ -1,21 +1,36 @@
+
+    @property
+    def is_scroll_locked(self):
+        return self.find_element(*self._scroll_lock_locator)
```
is there no way to actually try to scroll it? just looking for a css class isn't really testing this

Malar-Natarajan replied:
@TomWoodward I can try capturing the exception that raises, similar to verifying the disabled state of the elements

Malar-Natarajan replied:
@TomWoodward with latest commit, I have used scroll_from_element() and captured the exception in assert statement.
I tried scroll(x,y) but that would not throw any exception. 
Let me know if you think some other methods can be used more effectively here.

Malar-Natarajan replied:
@TomWoodward updated the test to use the random click as mentioned above

TomWoodward commented on pytest-selenium/pages/base.py:
```diff
@@ -34,13 +47,21 @@ def wait_for_region_to_display(self, region):
+            lambda _: title_before_click != (self.title.get_attribute("innerHTML") or "")
+        )
+
+    def assert_element_not_interactable_exception(self, element):
```
```suggestion
    def assert_element_not_interactable(self, element):
```

TomWoodward commented on pytest-selenium/pages/content.py:
```diff
@@ -36,6 +51,31 @@ def sidebar(self):
+        """
+        actionChains = ActionChains(self.driver)
+        actionChains.move_to_element_with_offset(self.driver.find_element_by_tag_name("body"), 0, 0)
+        actionChains.move_by_offset(439, 164).click().perform()
```
@m1yag1 is this the best way to test this? the hardcoded pixel values make me nervous

Malar-Natarajan replied:
@TomWoodward 
Mike suggested this as an alternative method instead of hardcoding the pixels:
1. Calculate the sidebar width/height & window width/height
2. Use random function to select a random (x, y) coordinate 
3. Use action chains to click on the (x,y) coordinate

Do you want me to update this test with this suggestion?

cc @m1yag1

Malar-Natarajan replied:
@TomWoodward updated the test to use the random click as mentioned above

TomWoodward commented on pytest-selenium/tests/test_toc.py:
```diff
@@ -64,3 +64,34 @@ def test_toc_toggle_button_opens_and_closes(selenium, base_url, book_slug, page_
+    # AND clicking anywhere in the content overlay should just close the TOC and content stays in the same page
+    content.click_content_overlay()
+    assert not sidebar.header.is_displayed
+    initial_url = base_url + "/books/" + book_slug + "/pages/" + page_slug
```
instead of calculating this url, can you save `selenium.current_url` to a variable before you start clicking on stuff?

TomWoodward commented on pytest-selenium/tests/test_toc.py:
```diff
@@ -64,3 +64,34 @@ def test_toc_toggle_button_opens_and_closes(selenium, base_url, book_slug, page_
+
+    # AND clicking anywhere in the content overlay should just close the TOC and content stays in the same page
+    content.click_content_overlay()
+    assert not sidebar.header.is_displayed
```
i feel like we could use a better helper for "sidebar is open or not" rather than going through the header

Malar-Natarajan replied:
@TomWoodward I just reused whatever we did in a previous TOC test (test_toc_toggle_button_opens_and_closes)

Malar-Natarajan replied:
@TomWoodward I updated the test to verify that just `sidebar` is displayed. The header was just capturing the 'X' button for mobile TOC

Malar-Natarajan committed:
Update pytest-selenium/pages/base.py

Co-Authored-By: Thomas Woodward <tom.g.woodward@gmail.com>

Malar-Natarajan committed:
updated tests with revieew comments

Malar-Natarajan committed:
Merge branch 'test-mobile-interaction' of https://github.com/openstax/rex-web into test-mobile-interaction

Malar-Natarajan committed:
Merge branch 'master' into test-mobile-interaction

Malar-Natarajan committed:
Merge branch 'master' into test-mobile-interaction

Malar-Natarajan committed:
resolving conflicts

Malar-Natarajan committed:
Merge branch 'test-mobile-interaction' of https://github.com/openstax/rex-web into test-mobile-interaction

Malar-Natarajan committed:
updated incorrect test

Malar-Natarajan commented:
@TomWoodward
Mike suggested this as an alternative method instead of hardcoding the pixels:

Calculate the sidebar width/height & window width/height
Use random function to select a random (x, y) coordinate
Use action chains to click on the (x,y) coordinate
Do you want me to update this test with this suggestion?

Malar-Natarajan commented:
@TomWoodward 
I am almost done with incorporating Mike's suggestion. I shall commit and re-submit for review

Malar-Natarajan committed:
Updated test:
use random() to generate a x,y and click/scroll in the content overlay instead of hardcoding pixel values

Malar-Natarajan committed:
Merge branch 'master' into test-mobile-interaction

Malar-Natarajan commented:
@TomWoodward This is ready for review

TomWoodward commented on pytest-selenium/pages/content.py:
```diff
@@ -50,6 +57,30 @@ def attribution(self):
     def section_url_within_attribution(self):
         return self.find_element(*self._section_url_locator)
 
+    def scroll_over_content_overlay(self, x, y):
```
since this function is very specifically for the content overlay, why are x and y not calculated within the function?

TomWoodward commented on pytest-selenium/pages/content.py:
```diff
@@ -50,6 +57,30 @@ def attribution(self):
+        touchActions = TouchActions(self.driver)
+        touchActions.scroll_from_element(self.print, x, y).perform()
+
+    def click_content_overlay(self, x, y):
```
same

Malar-Natarajan committed:
move area calculation from test

TomWoodward commented on pytest-selenium/tests/test_toc.py:
```diff
@@ -90,32 +88,32 @@ def test_local_toc_disables_interacting_with_content_on_mobile(
+    # sidebar_width = content.width(sidebar.root)
+    # window_width = content.window_width
+    # sidebar_width_left_offset = sidebar.root.get_attribute("offsetLeft")
+    # sidebar_width_offset = int(sidebar_width) + int(sidebar_width_left_offset)
```
commented code

Malar-Natarajan committed:
width/height calculation moved inside content.py

TomWoodward committed:
Merge branch 'master' into test-mobile-interaction
